File: vm/tests/src/limits.erl
%% Fixture for vm/tests/limits.rs: each function trips one resource limit and reports how the
%% VM ended the offender. Only BIFs are used: the tests load no OTP modules. Rebuild: erlc +deterministic -o vm/tests/fixtures vm/tests/src/limits.erl
-module(limits).
-export([mailbox/0, mailbox_self/0, heap/0, heap_flag/0, heap_spawn_opt/0, heap_ok/0,
         ets/0, memory/0]).

%% Another process floods a receiver that never reads.
mailbox() ->
    {Pid, Ref} = spawn_opt(fun() -> receive never -> ok end end, [monitor]),
    flood(Pid, 1000),
    receive {'DOWN', Ref, process, Pid, Reason} -> Reason after 1000 -> alive end.

%% A process floods itself; trapping exits does not save it.
mailbox_self() ->
    {Pid, Ref} = spawn_opt(fun() -> process_flag(trap_exit, true), flood(self(), 1000) end, [monitor]),
    receive {'DOWN', Ref, process, Pid, Reason} -> Reason after 1000 -> alive end.

flood(_, 0) -> ok;
flood(Pid, N) -> Pid ! {msg, N}, flood(Pid, N - 1).

%% Grows a list held live on the stack until the VM-wide limit kills it.
heap() ->
    {Pid, Ref} = spawn_opt(fun() -> grow([], 1 bsl 30) end, [monitor]),
    receive {'DOWN', Ref, process, Pid, Reason} -> Reason after 1000 -> alive end.

%% A process lowers its own limit.
heap_flag() ->
    {Pid, Ref} = spawn_opt(fun() ->
        process_flag(max_heap_size, #{size => 1000, kill => true}),
        grow([], 1 bsl 30) end, [monitor]),
    receive {'DOWN', Ref, process, Pid, Reason} -> Reason after 1000 -> alive end.

heap_spawn_opt() ->
    {Pid, Ref} = spawn_opt(fun() -> grow([], 1 bsl 30) end, [monitor, {max_heap_size, 1000}]),
    receive {'DOWN', Ref, process, Pid, Reason} -> Reason after 1000 -> alive end.

%% Below the limit nothing happens.
heap_ok() ->
    {Pid, Ref} = spawn_opt(fun() -> exit(length(grow([], 100))) end,
                           [monitor, {max_heap_size, 100000}]),
    receive {'DOWN', Ref, process, Pid, Reason} -> Reason after 1000 -> alive end.

grow(L, 0) -> L;
grow(L, N) -> grow([N | L], N - 1).

%% Fills a table until inserts fail.
ets() ->
    T = ets:new(t, [set]),
    try fill(T, 0) catch error:system_limit -> {system_limit, ets:info(T, size) > 0} end.

fill(T, N) -> ets:insert(T, {N, grow([], 100)}), fill(T, N + 1).

memory() ->
    Total = erlang:memory(total),
    Procs = erlang:memory(processes),
    {memory, M} = process_info(self(), memory),
    {is_integer(Total), Procs =< Total, M > 0}.

File: vm/tests/limits.rs
//! Resource limits: a process that exceeds one is ended the way BEAM would end it, and nothing
//! else is disturbed. BEAM has no limits for some of these, so these are not differential tests.
//! The fixture's source is `src/limits.erl`.

use std::collections::BTreeMap;

use beamlet_vm::platform::{Platform, PlatformError};
use beamlet_vm::vm::Limits;
use beamlet_vm::Vm;

struct TestPlatform {
    now: u64,
}

impl Platform for TestPlatform {
    fn monotonic_us(&mut self) -> u64 {
        self.now += 1;
        self.now
    }
    fn system_time_us(&mut self) -> Option<u64> {
        None
    }
    fn idle(&mut self, deadline: Option<u64>) {
        if let Some(d) = deadline {
            self.now = self.now.max(d);
        }
    }
    fn console_write(&mut self, _bytes: &[u8]) {}
    fn random(&mut self, _buf: &mut [u8]) -> Result<(), PlatformError> {
        Err(PlatformError::Unavailable)
    }
    fn load_module(&mut self, module: &str) -> Option<Vec<u8>> {
        let modules: BTreeMap<&str, &[u8]> =
            [("limits", include_bytes!("fixtures/limits.beam").as_slice())].into_iter().collect();
        modules.get(module).map(|b| b.to_vec())
    }
}

/// Run `limits:f()` under `limits` and return its result as text.
fn run(f: &str, limits: Limits) -> String {
    let mut vm = Vm::with_limits(Box::new(TestPlatform { now: 0 }), limits);
    let pid = vm.spawn("limits", f, |_| Vec::new()).unwrap();
    let r = vm.run_bounded(pid, 1_000_000).expect("finished");
    match r.unwrap() {
        Ok(t) => t.to_string(),
        Err(e) => format!("raised {}", e.reason),
    }
}

fn small() -> Limits {
    Limits { max_mailbox: 100, max_heap_words: 1 << 20, max_ets_words: 1 << 16, ..Limits::default() }
}

#[test]
fn full_mailbox_kills_the_receiver() {
    assert_eq!(run("mailbox", small()), "{system_limit,message_queue}");
}

#[test]
fn full_own_mailbox_kills_the_sender() {
    assert_eq!(run("mailbox_self", small()), "{system_limit,message_queue}");
}

#[test]
fn a_roomy_mailbox_is_not_a_limit() {
    let limits = Limits { max_mailbox: 10_000, ..small() };
    assert_eq!(run("mailbox", limits), "alive");
}

#[test]
fn the_vm_heap_limit_kills() {
    assert_eq!(run("heap", small()), "killed");
}

#[test]
fn a_process_can_lower_its_own_limit() {
    assert_eq!(run("heap_flag", Limits::default()), "killed");
}

#[test]
fn spawn_opt_sets_a_limit() {
    assert_eq!(run("heap_spawn_opt", Limits::default()), "killed");
}

#[test]
fn under_the_limit_nothing_happens() {
    assert_eq!(run("heap_ok", small()), "100");
}

#[test]
fn ets_inserts_past_the_limit_raise() {
    assert_eq!(run("ets", small()), "{system_limit,true}");
}

#[test]
fn memory_is_reported() {
    assert_eq!(run("memory", small()), "{true,true,true}");
}

#[test]
fn programs_need_the_platform_to_grant_them() {
    assert_eq!(run("no_programs", small()), "{error,eacces}");
}
